userspace/libscap/scap_procs.c

```diff
@@ -561,6 +564,55 @@ int32_t scap_proc_fill_loginuid(scap_t *handle, struct scap_threadinfo* tinfo, c
 	}
 }
 
+int32_t scap_proc_fill_exe_writable(scap_t* handle, struct scap_threadinfo* tinfo,  uint32_t uid, uint32_t gid, const char *procdirname, const char *exetarget)
+{
+	char proc_exe_path[SCAP_MAX_PATH_SIZE];
+	struct stat targetstat;
+
+	snprintf(proc_exe_path, sizeof(proc_exe_path), "%sroot%s", procdirname, exetarget);
+
+	// if the file doesn't exist we can't determine if it was writable, assume false
+	if(stat(proc_exe_path, &targetstat) < 0)
+	{
+		return SCAP_SUCCESS;
+	}
+
+	// if you're the user owning the file you can chmod, so you can effectively write to it
+	if(targetstat.st_uid == uid) {
+		tinfo->exe_writable = true;
+		return SCAP_SUCCESS;
+	}
+
+	int orig_uid = setfsuid(uid);
+	int orig_gid = setfsgid(gid);
+	if(setfsuid(-1) == uid && setfsgid(-1) == gid) {
+		int fd = open(proc_exe_path, O_WRONLY);
```

Comment:
access/faccess(at) instead of opening the file for write?

Reply:
I believe I considered that but AFAICT `access` uses the real UID/GID which in case of Falco is always root. Do you know if any variant of `access` uses the fsuid/fsgid to perform the check?

Reply:
However, `euidaccess()` exists and can be used with `euid`/`egid`. One clear advantage of doing that instead of `open()` is that `open()` might return other errors, such as file busy.